KAFKA-5856: Add AdminClient.createPartitions()

tombentley opened the pull request:
See KIP-195.

The contribution is my original work and I license the work to the project under the project's open source license.

This patch adds AdminClient.createPartitions() and the network protocol is
uses. The broker-side algorithm is as follows:

1. KafkaApis makes some initial checks on the request, then delegates to the
   new AdminManager.createPartitions() method.
2. AdminManager.createPartitions() performs some validation then delegates to
   AdminUtils.addPartitions().

Aside: I felt it was safer to add the extra validation in
AdminManager.createPartitions() than in AdminUtils.addPartitions() since the
latter is used on other code paths which might fail differently with the
introduction of extra checks.

3. AdminUtils.addPartitions() does its own checks and adds the partitions.
4. AdminManager then uses the existing topic purgatory to wait for the
   PartitionInfo available from the metadata cache to become consistent with
   the new total number of partitions.

The messages of exceptions thrown in AdminUtils affecting this new API have
been made consistent with initial capital letter and terminating period.
A few have been reworded for clarity. I've also standardized on using
String.format().

cc @ijuma

tombentley committed:
KAFKA-5856: Add AdminClient.createPartitions()

See KIP-195.

This patch adds AdminClient.createPartitions() and the network protocol is
uses. The broker-side algorithm is as follows:

1. KafkaApis makes some initial checks on the request, then delegates to the
   new AdminManager.createPartitions() method.
2. AdminManager.createPartitions() performs some validation then delegates to
   AdminUtils.addPartitions().

Aside: I felt it was safer to add the extra validation in
AdminManager.createPartitions() than in AdminUtils.addPartitions() since the
latter is used on other code paths which might fail differently with the
introduction of extra checks.

3. AdminUtils.addPartitions() does its own checks and adds the partitions.
4. AdminManager then uses the existing topic purgatory to wait for the
   PartitionInfo available from the metadata cache to become consistent with
   the new total number of partitions.

The messages of exceptions thrown in AdminUtils affecting this new API have
been made consistent with initial capital letter and terminating period.
A few have been reworded for clarity. I've also standardized on using
String.format().

tedyu commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1799,4 +1801,42 @@ void handleFailure(Throwable throwable) {
+                completeAllExceptionally(futures.values(), throwable);
+            }
+        }, now);
+        return new CreatePartitionsResult(new HashMap<String, KafkaFuture<Void>>(futures));
```
Another HashMap is created.
Is it possible not to ?

tombentley replied:
@tedyu without the copy `futures` would be simultaneously accessible to both the `Call` and the `CreatePartitionsResult`. I will admit I didn't think very hard about it because the same pattern is used by all the other methods in `KafkaAdminClient`. 

As it happens both `Call` and `CreatePartitionsResult` only call `get()` and `values()` on the `HashMap`, so if you're willing it make assumptions about the internals of `HashMap`, you could make the argument that it would be safe for them to share the instance. But making assumptions and correct multi-threaded code have plenty of ugly history, so I think making a copy is warranted on the grounds of defensive programming.

tedyu replied:
When we get here, futures Map is already filled.
Calling get() and values() does not involve the internals of HashMap when we consider thread-safety.
What do you think ?

tedyu replied:
See KAFKA-5911

ijuma replied:
I agree that it's safe to avoid the copy here. By making the copy, we are iterating from a different thread anyway, so it doesn't add any safety.

tedyu commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1799,4 +1801,42 @@ void handleFailure(Throwable throwable) {
+        for (String topic: newPartitions.keySet()) {
+            futures.put(topic, new KafkaFutureImpl<Void>());
+        }
+        final Map<String, NewPartition> requestMap = new HashMap<>(newPartitions);
```
Can newPartitions be passed directly (without creating requestMap) ?

tombentley replied:
No, because the `newPartitions` is the one passed by the caller who could modify it while it's being concurrently accessed by the `Call`.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/CreatePartitionsOptions.java:
```diff
@@ -0,0 +1,51 @@
+    /**
+     * Set to true if the request should be validated without creating new partitions.
+     */
+    public CreatePartitionsOptions shouldValidateOnly(boolean validateOnly) {
```
`shouldValidateOnly` => `validateOnly` for consistency with other options

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/NewPartition.java:
```diff
@@ -0,0 +1,99 @@
+ * The API of this class is evolving, see {@link AdminClient} for details.
+ */
+@InterfaceStability.Evolving
+public class NewPartition {
```
This should be plural `NewPartitions`? It does say `NewPartitions` in the KIP.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/admin/NewPartition.java:
```diff
@@ -0,0 +1,99 @@
+     * @param newCount         The new total partition count (not the number of new partitions).
+     * @param newAssignments   The replica assignments for the new partitions.
+     */
+    public static NewPartition increaseTo(int newCount, List<List<Integer>> newAssignments) {
```
Can we improve on the names of these parameters? One applies to `new` and other applies to new & existing, but that is impossible to tell from the names. Perhaps `total` is better than `count`, or use a different prefix instead of `new` for count.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/protocol/Protocol.java:
```diff
@@ -1895,6 +1895,30 @@
+            new Field("topic_partition_count", new ArrayOf(
+                    new Schema(
+                            new Field("topic", STRING, "The topic name."),
+                            new Field("partition_count", new Schema(
```
The name `partition_count` doesn't quite seem right since it is a combination of count and assignments.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/CreatePartitionsRequest.java:
```diff
@@ -0,0 +1,198 @@
+                    int j = 0;
+                    for (Integer broker : partitionAssignment) {
+                        x[i][j] = broker;
+                        j++;
```
Isn't this the same as `partitionAssignment.toArray()`?

rajinisivaram commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -376,6 +378,41 @@ public void testDescribeConfigs() throws Exception {
         }
     }
 
+    public void testAlterPartitionCounts() throws Exception {
```
Needs `@Test` annotation

rajinisivaram commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -376,6 +378,41 @@ public void testDescribeConfigs() throws Exception {
+            CreatePartitionsResult results = env.adminClient().createPartitions(counts);
+            Map<String, KafkaFuture<Void>> values = results.values();
+            KafkaFuture<Void> myTopicResult = values.get("my_topic");
+            assertTrue(!myTopicResult.isCompletedExceptionally());
```
Add error message for `assertTrue`

rajinisivaram commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -376,6 +378,41 @@ public void testDescribeConfigs() throws Exception {
+            assertTrue(!myTopicResult.isCompletedExceptionally());
+            assertEquals(null, myTopicResult.get());
+            KafkaFuture<Void> otherTopicResult = values.get("other_topic");
+            //assertTrue(otherTopicResult.isCompletedExceptionally());
```
This should be enabled after the `get`?

tombentley replied:
I removed this assertion. The problem was that it would fail because the future hadn't completed (exceptionally or not): IWO `isCompletedExceptionally()` doesn't wait if the future isn't completed yet. The `get()` call throwing or not is a sufficient test of the functionality I'm adding here.

rajinisivaram commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -376,6 +378,41 @@ public void testDescribeConfigs() throws Exception {
+            //assertTrue(otherTopicResult.isCompletedExceptionally());
+            try {
+                otherTopicResult.get();
+                fail();
```
Add error message to `fail`

rajinisivaram commented on core/src/main/scala/kafka/admin/AdminUtils.scala:
```diff
@@ -134,17 +134,17 @@ object AdminUtils extends Logging with AdminUtilities {
+      throw new InvalidReplicationFactorException("Replication factor must be larger than 0.")
     if (replicationFactor > brokerMetadatas.size)
-      throw new InvalidReplicationFactorException(s"replication factor: $replicationFactor larger than available brokers: ${brokerMetadatas.size}")
+      throw new InvalidReplicationFactorException("Replication factor: %s larger than available brokers: %s.".format(replicationFactor, brokerMetadatas.size))
```
Any reason the string interpolation was replaced here?

ijuma replied:
Yes, we should favour String interpolation in Scala code. It's both significantly faster and safer.

tombentley replied:
np, there was already a mixture in the code and I standardised on the wrong one.

rajinisivaram commented on core/src/main/scala/kafka/admin/AdminUtils.scala:
```diff
@@ -272,18 +272,25 @@ object AdminUtils extends Logging with AdminUtilities {
     val partitionsToAdd = numPartitions - existingPartitionsReplicaList.size
     if (partitionsToAdd <= 0)
-      throw new AdminOperationException("The number of partitions for a topic can only be increased")
+      throw new InvalidPartitionsException("The number of partitions for a topic can only be increased, topic %s currently has %d partitions.".format(topic, existingPartitionsReplicaList.size))
```
Use string interpolation? There are a few cases where format can be replaced.

rajinisivaram commented on core/src/main/scala/kafka/admin/AdminUtils.scala:
```diff
@@ -272,18 +272,25 @@ object AdminUtils extends Logging with AdminUtilities {
     val existingPartitionsReplicaList = zkUtils.getReplicaAssignmentForTopics(List(topic))
     if (existingPartitionsReplicaList.isEmpty)
-      throw new AdminOperationException("The topic %s does not exist".format(topic))
+      throw new InvalidTopicException("The topic %s does not exist".format(topic))
```
Use string interpolation instead of format?

rajinisivaram commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -40,6 +42,65 @@ class AdminManager(val config: KafkaConfig,
                    val metadataCache: MetadataCache,
                    val zkUtils: ZkUtils) extends Logging with KafkaMetricsGroup {
+
+  def createPartitions(timeout: Int,
```
Can this method be moved later in this class - after field declaration etc.? Probably after `createTopics`?

ijuma replied:
Yes, because variable initialisation order matters (unlike `defs`), it's best to have them all before the methods.

rajinisivaram commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -360,6 +360,133 @@ class AdminClientIntegrationTest extends KafkaServerTestHarness with Logging {
+    TestUtils.createTopic(zkUtils, topic1, 1, 1, servers, new Properties)
+
+    // assert that the topic has 1 partition
+    assertEquals(1, client.describeTopics(Set(topic1).asJavaCollection).values().get(topic1).get().partitions().size())
```
`()` can be removed from `values()`, `partitions()` etc. Many cases of this in the method.

rajinisivaram commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -360,6 +360,133 @@ class AdminClientIntegrationTest extends KafkaServerTestHarness with Logging {
+    // assert that a validateOnly request doesn't increase the number of partitions
+    var alterResult = client.createPartitions(Map(topic1 -> NewPartition.increaseTo(new Integer(2))).asJava, validateOnly)
+    var altered = alterResult.values.get(topic1).get
+    // assert that the topics now has 2 partitions
```
Comment should say `1 partition`?

rajinisivaram reviewed (commented):
@tombentley Thank you for the PR. I have left a few minor comments.

rajinisivaram commented on core/src/main/scala/kafka/server/DelayedCreatePartitions.scala:
```diff
@@ -0,0 +1,93 @@
+  extends DelayedOperation(delayMs) {
+
+  // the results of topics we were waiting for
+  var waitedResults = new mutable.HashMap[String, ApiError]();
```
Should be `val` rather than `var`? Also don't need semicolon here and other places in the scala code.
The initialization could be `mutable.Map.empty[String, ApiError]`.

tombentley commented:
Thanks for the review @rajinisivaram, @ijuma, sorry about the java-isms. Those are all fixed now.

tombentley commented:
retest this please

ijuma commented:
@tombentley, can you please fix the conflicts in this branch? As you saw, we have moved the protocol definition to the request/response classes in https://github.com/apache/kafka/commit/0cf7708007b01faac5012d939f3c50db274f858d

rajinisivaram commented on clients/src/main/java/org/apache/kafka/common/requests/CreatePartitionsRequest.java:
```diff
@@ -0,0 +1,193 @@
+    @Override
+    protected Struct toStruct() {
+        Struct struct = new Struct(ApiKeys.CREATE_PARTITIONS.requestSchema(version()));
+        List<Struct> topicParitionsList = new ArrayList<>();
```
typo : topicParitionsList => topicPartitionsList

rajinisivaram commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -194,6 +197,64 @@ class AdminManager(val config: KafkaConfig,
+          case Some(count) => count
+          case None => throw new InvalidTopicException()
+        }
+        val newNumPartitions = newPartition.totalCount()
```
() not required for `totalCount()` and other getters like `assignments()` below.

rajinisivaram commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -194,6 +197,64 @@ class AdminManager(val config: KafkaConfig,
+        if (numPartitionsIncrement <= 0) {
+          throw new InvalidPartitionsException(s"Topic currently has $oldNumPartitions partitions: Cannot create $numPartitionsIncrement partitions.")
+        }
+        val reassignmentStr = if (newPartition.assignments() == null) {
```
You can use `match` instead of `if`:
`newPartition.assignments match {
case null => ""
case assignments => ...
}

ijuma replied:
In cases like this, either option is fine. I even prefer `if/else` since it's more concise. Pattern matching is beneficial when it's used to extract data, but that is not happening here.

rajinisivaram replied:
@ijuma is the scala expert, so please leave as is :-)

rajinisivaram commented on core/src/main/scala/kafka/server/DelayedCreatePartitions.scala:
```diff
@@ -0,0 +1,93 @@
+  val waitedResults = mutable.Map.empty[String, ApiError]
+
+  // the topics we're still waiting for
+  val waitingFor = new mutable.HashMap[String, NewPartitions]() ++= waiting
```
mutable.Map.empty as in the previous statement

rajinisivaram commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1333,6 +1335,40 @@ class KafkaApis(val requestChannel: RequestChannel,
+    } else {
+      // Special handling to add duplicate topics to the response
+      val dupes = alterPartitionCountsRequest.duplicates().asScala
+      val notDuped = alterPartitionCountsRequest.newPartitions.asScala.retain((topic, count)=> !dupes.contains(topic))
```
space before =>

rajinisivaram commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1333,6 +1335,40 @@ class KafkaApis(val requestChannel: RequestChannel,
+      // Special handling to add duplicate topics to the response
+      val dupes = alterPartitionCountsRequest.duplicates().asScala
+      val notDuped = alterPartitionCountsRequest.newPartitions.asScala.retain((topic, count)=> !dupes.contains(topic))
+      val (authorized, unauthorised) = notDuped.partition { case (topic, _) =>
```
unauthorised => unauthorized?

rajinisivaram commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -360,6 +360,133 @@ class AdminClientIntegrationTest extends KafkaServerTestHarness with Logging {
+      fail("Expect InvalidPartitionsException when #brokers != replication factor")
+    } catch {
+      case e: ExecutionException =>
+        assertTrue(e.getCause.isInstanceOf[InvalidReplicaAssignmentException])//TODO KIP doesn't say this
```
Can you update the KIP and remove the comment?

rajinisivaram commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -360,6 +360,133 @@ class AdminClientIntegrationTest extends KafkaServerTestHarness with Logging {
+    }
+
+    // try with empty assignments
+    alterResult = client.createPartitions(Map(topic1 -> NewPartitions.increaseTo(new Integer(4), Arrays.asList())).asJava, validateOnly)
```
Arrays.asList() => `Collections.emptyList`

rajinisivaram reviewed (commented):
@tombentley Thank you for the updates. Left a few more minor comments. Once the conflicts are resolved and the comments addressed, I can merge this.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -417,4 +417,33 @@ public DescribeReplicaLogDirResult describeReplicaLogDir(Collection<TopicPartiti
+     * @param options       The options to use when creating the new paritions.
+     * @return              The CreatePartitionsResult.
+     */
+    public abstract CreatePartitionsResult createPartitions(Map<String, NewPartitions> newPartitions,
```
We should add a blurb similar to `createTopics` with regards to the delay of metadata propagation.

Also, we should mention in the other overload that it's a convenience method and point to this for all details. Seems like we have not done that in the `createTopics` convenience method and it would be good to fix it as part of this PR as well.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/admin/NewPartitions.java:
```diff
@@ -0,0 +1,99 @@
+     * Increase the partition count for a topic to the given {@code totalCount}.
+     * The assignment of new replicas to brokers will be decided by the broker.
+     *
+     * @param totalCount       The new total partition count (not the number of new partitions).
```
Nit: no need to align across methods as it wastes space for not much benefit. I think some of the AdminClient classes do this, but it's not part of the Kafka style guide.

Maybe "The total partitions count after the operation succeeds" or something like that. Same for the other method.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/admin/NewPartitions.java:
```diff
@@ -0,0 +1,99 @@
+     *
+     * <p>For example, suppose a topic currently has a replication factor of 2, and
+     * has 3 partitions. The number of partitions can be increased to 4
+     * (with broker 1 being the preferred replica for the new partition) using a
```
Maybe we should use an example where 2 partitions are added. Also, we can assume that `Arrays.asList` has been statically imported to keep it concise.

tombentley replied:
If the example shows two partitions being added then there's more possibility for confusion because the replication factor is also 2, so the list of lists is 2×2 and it's less obvious how the size of the list of lists is related to the number of additional partitions and the replication factor. So I'll add three partitions in the example.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/admin/NewPartitions.java:
```diff
@@ -0,0 +1,99 @@
+        bld.append("(totalCount=").append(totalCount()).
+        append(", newAssignments=").append(assignments()).
+        append(")");
+        return bld.toString();
```
There isn't a benefit in using `StringBuilder` in this case, so may as well use String concat for less noise.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/admin/NewPartitions.java:
```diff
@@ -0,0 +1,99 @@
+
+    private NewPartitions(int totalCount, List<List<Integer>> newAssignments) {
+        this.totalCount = totalCount;
+        this.newAssignments = newAssignments;
```
We should validate the parameters and we should have tests for that.

tombentley replied:
I tried putting validation here. Obviously we can't validate properly because we don't know the replication factor or existing number of partitions.

We still need to have the server-side code for validating these things (non-Java clients), so I reflectively created invalid NewPartitions in the `AdminClientIntegrationTest` in order to test that. But that reveals a new problem: The validation in `NewPartitions` applies during deserialization of the `CreatePartitionsRequest`, which prevents the request making it as far as `KafkaApis` and `AdminManager` where "proper" validation (with knowledge of the current number of partitions, replication factor etc) occurs. Instead, exceptions propagating during deserialization seem to lead to the broker disconnecting from the client, presumably because it considers an exception thrown during deserialization to be an unrecoverable client error.

I'm minded to say we should just let the server do the validation, but maybe you have a better idea, @ijuma?

ijuma replied:
That's a good point about validation causing issues when the broker tries to deserialize. I was just thinking that we'd validate that `newAssignments` was consistent, i.e. every inner list has the same size. Maybe we should do it in the static factory methods which are only used by the `AdminClient` instead of the broker.

tombentley replied:
Doing it in the factory methods doesn't solve the problem though, because if we allow the broker a way to create a `NewPartitions` that avoids the validation, we necessarily expose that same mechanism to the client, defeating the point of having the validation in `NewPartitions`.

ijuma replied:
Client-side validation never guarantees anything (people can always bypass them), this check would be simply to fail-fast in case of programming mistakes. In any case, if you think the benefit is not worth the extra complexity, we can leave it out.

ijuma commented on clients/src/main/java/org/apache/kafka/common/errors/ReassignmentInProgressException.java:
```diff
@@ -0,0 +1,28 @@
+ */
+public class ReassignmentInProgressException extends ApiException {
+
+    public ReassignmentInProgressException(String msg) {
```
We should probably have a constructor that takes a message and an exception as well.

ijuma commented on clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java:
```diff
@@ -167,8 +169,9 @@ public Struct parseResponse(short version, ByteBuffer buffer) {
-            SaslAuthenticateResponse.schemaVersions());
-
+            SaslAuthenticateResponse.schemaVersions()),
+    CREATE_PARTITIONS(37, "CreatePartitions",CreatePartitionsRequest.schemaVersions(),
```
Nit: space missing after `,`.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/CreatePartitionsRequest.java:
```diff
@@ -0,0 +1,218 @@
+    private static final String TIMEOUT_KEY_NAME = "timeout";
+    private static final String VALIDATE_ONLY_KEY_NAME = "validate_only";
+
+    public static final Schema CREATE_PARTITIONS_REQUEST_V0 = new Schema(
```
Does this need to be public?

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/CreatePartitionsRequest.java:
```diff
@@ -0,0 +1,218 @@
+                            )))),
+                    "List of topic and the corresponding partition count"),
+            new Field("timeout", INT32, "The time in ms to wait for a topic to be altered."),
+            new Field("validate_only", BOOLEAN, "If true then validate the request, but don't actually increase the partition count."));
```
We can refer to the constants instead of hardcoding the field names as strings.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/CreatePartitionsRequest.java:
```diff
@@ -0,0 +1,218 @@
+    private final int timeout;
+    private final boolean validateOnly;
+
+
```
Extra empty lines?

ijuma reviewed (commented):
Thanks for the PR, I left some initial comments. I think they should all be easy to address. I'll take another pass a bit later. From what I've seen, I think we're good to get this into 1.0.0.

ijuma commented:
Also, there are checkstyle errors.

ijuma commented:
Btw, for future changes to this PR, please add them as separate commits. That makes it much easier to review. The commits get squashed as part of the merge either way.

tombentley commented:
No problem. I just saw that PRs got merged as a single commit and assumed that was the standard.

tombentley committed:
Fix review comments and test some more cases of invalid requests

As a result of the extra tests I have revised the exception messages slightly.

tombentley commented:
@ijuma all your comments should have been addressed

ijuma commented on clients/src/main/java/org/apache/kafka/clients/admin/CreatePartitionsResult.java:
```diff
@@ -0,0 +1,53 @@
+     * Return a future which succeeds if all the partition creations succeed.
+     */
+    public KafkaFuture<Void> all() {
+        return KafkaFuture.allOf(values.values().toArray(new KafkaFuture[values.size()]));
```
Nit: there's no need to pass the size to the array. Somewhat surprisingly, it's actually slower than passing `0` (and more verbose too).

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/CreatePartitionsRequest.java:
```diff
@@ -0,0 +1,215 @@
+                            )))),
+                    "List of topic and the corresponding partition count"),
+            new Field(TIMEOUT_KEY_NAME, INT32, "The time in ms to wait for a topic to be altered."),
+            new Field(VALIDATE_ONLY_KEY_NAME, BOOLEAN, "If true then validate the request, but don't actually increase the partition count."));
```
Nit: good to avoid lines longer than GitHub's review window.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/CreatePartitionsRequest.java:
```diff
@@ -0,0 +1,215 @@
+                                    new Field(ASSIGNMENT_KEY_NAME, ArrayOf.nullable(new ArrayOf(INT32)), "The assigned brokers")
+                            )))),
+                    "List of topic and the corresponding partition count"),
+            new Field(TIMEOUT_KEY_NAME, INT32, "The time in ms to wait for a topic to be altered."),
```
Nit: "topic to be altered" -> "partitions created"?

ijuma commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -378,6 +380,40 @@ public void testDescribeConfigs() throws Exception {
+            CreatePartitionsResult results = env.adminClient().createPartitions(counts);
+            Map<String, KafkaFuture<Void>> values = results.values();
+            KafkaFuture<Void> myTopicResult = values.get("my_topic");
+            assertEquals(null, myTopicResult.get());
```
Since this is a `KafkaFuture<Void>`, I think it's actually clearer to just do the `get()` without any assert (we are doing it purely for the side-effect).

ijuma commented on core/src/main/scala/kafka/admin/AdminUtils.scala:
```diff
@@ -25,7 +25,7 @@ import java.util.Random
 
 import kafka.common.TopicAlreadyMarkedForDeletionException
-import org.apache.kafka.common.errors.{InvalidPartitionsException, InvalidReplicaAssignmentException, InvalidReplicationFactorException, InvalidTopicException, TopicExistsException, UnknownTopicOrPartitionException}
+import org.apache.kafka.common.errors._
```
Was this intentional or was it changed by the IDE?

tombentley replied:
IDE, but I think it was triggered because it had to add `ReassignmentInProgressException` to the imports. Since I don't see it [mentioned](https://kafka.apache.org/coding-guide), what's the policy wrt wildcard imports?

ijuma replied:
We disallow them in Java via checkstyle:

https://github.com/apache/kafka/blob/trunk/checkstyle/checkstyle.xml#L72

Same policy for Scala, but not enforced, which results in inconsistency.

ijuma replied:
An exception to this is when using wildcard imports for the type class pattern. For example, `import scala.collection.JavaConverters._`.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java:
```diff
@@ -417,4 +419,40 @@ public DescribeReplicaLogDirResult describeReplicaLogDir(Collection<TopicPartiti
+     * Increase the number of partitions of the topics given as the keys of {@code newPartitions}
+     * according to the corresponding values.
+     *
+     * This is a convenience method for {@link #createPartitions(Map, CreatePartitionsOptions)} with default options. See the overload for more details.
```
Nit: long line.

ijuma commented on clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:
```diff
@@ -1075,4 +1083,22 @@ private AlterConfigsResponse createAlterConfigsResponse() {
+        return new CreatePartitionsRequest(assignments, 0, false, (short) 0);
+    }
+
+    private CreatePartitionsResponse createCreatePartitionsErrorResponse() {
```
Hmm, not sure why we are only testing the error case. Seems like we should create a response with both success and error.

ijuma commented on clients/src/test/java/org/apache/kafka/common/requests/RequestResponseTest.java:
```diff
@@ -1075,4 +1083,22 @@ private AlterConfigsResponse createAlterConfigsResponse() {
+        return new CreatePartitionsRequest(assignments, 0, false, (short) 0);
+    }
+
+    private CreatePartitionsRequest createCreatePartitionsRequestWithAssignments() {
```
Would be good to have at least 2 entries in this method and above?

ijuma commented on core/src/main/scala/kafka/admin/AdminUtils.scala:
```diff
@@ -272,18 +272,25 @@ object AdminUtils extends Logging with AdminUtilities {
+    if (zkUtils.pathExists(ZkUtils.ReassignPartitionsPath)) {
+      // We prevent addition partitions while a reassignment is in progress, since
+      // during reassignment there is no meaningful notion of replication factor
+      throw new ReassignmentInProgressException("A partition reassignment is in progress.")
```
If the goal is to ensure that the replication factor is sensible, then we should do this check after we get the replica assignment from ZK.

tombentley replied:
Why does that matter?

ijuma replied:
In theory, a reassignment could start right after we do this check and before we retrieve the data from ZK. Although it's unlikely for the number of replicas to be changed quickly enough, but generally good to avoid race conditions altogether.

ijuma commented on core/src/main/scala/kafka/admin/AdminUtils.scala:
```diff
@@ -272,18 +272,25 @@ object AdminUtils extends Logging with AdminUtilities {
 
     val existingReplicaListForPartitionZero = existingPartitionsReplicaList.find(p => p._1.partition == 0) match {
-      case None => throw new AdminOperationException("the topic does not have partition with id 0, it should never happen")
+      case None => throw new AdminOperationException("The topic does not have partition with id 0, it should never happen.")
```
Maybe we can rewrite this along the lines of s"Unexpected replica assignment for topic '$topic', partition id 0 is missing. Assignment: $existingPartitionsReplicaList".

ijuma commented on core/src/main/scala/kafka/admin/AdminUtils.scala:
```diff
@@ -272,18 +272,25 @@ object AdminUtils extends Logging with AdminUtilities {
     val partitionsToAdd = numPartitions - existingPartitionsReplicaList.size
     if (partitionsToAdd <= 0)
-      throw new AdminOperationException("The number of partitions for a topic can only be increased")
+      throw new InvalidPartitionsException(s"The number of partitions for a topic can only be increased, topic $topic currently has ${existingPartitionsReplicaList.size} partitions.")
```
We should include `numPartitions` in the message as that may be helpful in some cases. Also, long line.

ijuma commented on core/src/main/scala/kafka/admin/AdminUtils.scala:
```diff
@@ -272,18 +272,25 @@ object AdminUtils extends Logging with AdminUtilities {
+      throw new ReassignmentInProgressException("A partition reassignment is in progress.")
+    }
+
     val existingPartitionsReplicaList = zkUtils.getReplicaAssignmentForTopics(List(topic))
```
Maybe rename this to the more concise `existingAssignment`. Similar for other unnecessarily long variable names.

ijuma commented on core/src/main/scala/kafka/admin/AdminUtils.scala:
```diff
@@ -316,17 +323,16 @@ object AdminUtils extends Logging with AdminUtilities {
+      val brokerList = partitionList(i).split(":").map(s => s.trim().toInt).toList
       if (brokerList.isEmpty)
-        throw new AdminOperationException("replication factor must be larger than 0")
+        throw new InvalidReplicaAssignmentException("Replication factor must be larger than 0.")
```
We should include the partition id in the message. Same for every other error message.

ijuma commented on core/src/main/scala/kafka/admin/AdminUtils.scala:
```diff
@@ -316,17 +323,16 @@ object AdminUtils extends Logging with AdminUtilities {
-      ret.put(partitionId, brokerList.toList)
-      if (ret(partitionId).size != ret(startPartitionId).size)
-        throw new AdminOperationException("partition " + i + " has different replication factor: " + brokerList)
+        throw new AdminOperationException(s"Some specified brokers not available. Specified brokers: ${brokerList.mkString(", ")}, available brokers: ${availableBrokerList.mkString(", ")}.")
```
Why are we not converting this exception?

ijuma commented on core/src/main/scala/kafka/admin/AdminUtils.scala:
```diff
@@ -316,17 +323,16 @@ object AdminUtils extends Logging with AdminUtilities {
+        throw new AdminOperationException(s"Some specified brokers not available. Specified brokers: ${brokerList.mkString(", ")}, available brokers: ${availableBrokerList.mkString(", ")}.")
+      ret.put(partitionId, brokerList)
+      if (brokerList.size != ret(startPartitionId).size)
+        throw new InvalidReplicaAssignmentException(s"Partition ${i+startPartitionId} has different replication factor: ${brokerList.mkString(", ")}.")
```
We should probably have a separate method for validating this and provide a more useful message. Maybe something along the lines of "Inconsistent replication factor between partitions, partition 0 has X while partitions a, b, c have different values". Seems like we're doing something like this in `addPartitions`.

ijuma commented on core/src/main/scala/kafka/admin/AdminUtils.scala:
```diff
@@ -300,14 +307,14 @@ object AdminUtils extends Logging with AdminUtilities {
+      throw new InvalidReplicaAssignmentException(s"Existing topic replication factor of ${existingReplicaListForPartitionZero.size}, but manual replication assignment would imply replication factor(s) of ${unmatchedRepFactorList.map(_.size).mkString(",")}.")
+    if (!validateOnly) {
+      info(s"Add partition list for $topic is $newPartitionReplicaList.")
+      val partitionReplicaList = existingPartitionsReplicaList.map(p => p._1.partition -> p._2)
```
Maybe we can take the chance and clean-up this code a bit. In general `_1` and `_2` are not very readable. The following is clearer:

```scala
existingPartitionsReplicaList.map { case (tp, replicas) => (tp.partition, replicas) }
```

ijuma commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -194,6 +197,64 @@ class AdminManager(val config: KafkaConfig,
+        val oldNumPartitions = zkUtils.getTopicPartitionCount(topic) match {
+          case Some(count) => count
+          case None => throw new InvalidTopicException()
+        }
```
Nit: `getOrElse` is a more concise version of this. Also, shouldn't this be `UnknownTopicOrPartition`?

tombentley replied:
What is the purpose of `InvalidTopicException` then? Tell me and I'll update its javadoc.

ijuma replied:
It's typically used for cases where the topic name is invalid. For example, creating a topic with a name that is not allowed. Or trying an operation on an internal topic.

ijuma replied:
Btw, one difference is that `UnknownTopicOrPartitions` is retriable while `InvalidTopic` is not. So we should think about the impact of that. If someone creates a topic and soon after tries to add partitions, `UnknownTopicOrPartitions` would work better. However, for the common case, it seems that one would not want to retry if the topic doesn't exist.

ijuma commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -194,6 +197,64 @@ class AdminManager(val config: KafkaConfig,
+    var waiting = Map.empty[String, NewPartitions]
+    var done = Map.empty[String, ApiError]
+
+    newPartitions.foreach{case (topic, newPartition: NewPartitions) =>
```
Nit: space missing before `case`. Also, is there a reason why you're specifying the type for one of the parameters?

ijuma commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -194,6 +197,64 @@ class AdminManager(val config: KafkaConfig,
+        val newNumPartitions = newPartition.totalCount
+        val numPartitionsIncrement = newNumPartitions - oldNumPartitions
+        if (numPartitionsIncrement <= 0) {
+          throw new InvalidPartitionsException(s"Topic currently has $oldNumPartitions partitions: Cannot create $numPartitionsIncrement partitions.")
```
Maybe this should say something like s"Topic currently has $oldNumPartitions, which is higher than requested $newNumPartitions"?

ijuma commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -194,6 +197,64 @@ class AdminManager(val config: KafkaConfig,
+          if (assignments.size != numPartitionsIncrement) {
+            throw new InvalidRequestException(s"Increasing the number of partitions by $numPartitionsIncrement but ${assignments.size} assignments provided.")
+          }
+          "," * (oldNumPartitions) + assignments.map(_.mkString(":")).mkString(",")
```
This is not ideal. It's better to have the building block method accept an assignment in a structured form (i.e. not a string).

tombentley replied:
This is difficult to refactor: Currently when invoking `kafka-topics.sh --alter --partitiions ... --replica-assignment ...` we expect a complete replica assignment, but we ignore (not even validating) that part of the replica assignment string which is for the existing partitions. If I were to provide a `parseReplicaAssignment()` function and have `addPartitions()` take a `Map[Int, Seq[Int]]` then either `TopicCommand` or `parseReplicaAssignment()` would have to look up the existing number of partitions for itself to know which are the new assignments in the string. But `addPartitions()` also needs this info, so we consult ZK twice for the same data and the first lookup reintroduces the race wrt `zkUtils.pathExists(ZkUtils.ReassignPartitionsPath)`.

I suppose I can make `addPartitions()` take the complete assignment, rather than the assignment just for the replicas being added, but this retains a poor semantics of passing around information that isn't being used for anything, nor validated.

It might be better to leave it as it is currently and change the `TopicCommand` (which, I understand, is being rewritten anyway) to use a different `--new-replica-assignment` option with `--alter`, retaining the existing `--replica-assignment` for use with `--alter`, thus removing the source of the whole "pass around the complete replica map" issue. Note there are other issues with passing around the complete replica map: It makes it look like `kafka-topics.sh` can be used to reassign existing replicas.

wdyt?

ijuma replied:
What if one of the parameters of `addPartitions` was the current replica assignment? That would avoid the double look-up right?

tombentley replied:
That works, but it does mean we have to trust our caller that the existing assignments are indeed the correct ones, because we use them to update ZK. I've implemented it like this for now.

ijuma commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -18,18 +18,20 @@ package kafka.server
 import org.apache.kafka.common.protocol.Errors
 import org.apache.kafka.common.requests.CreateTopicsRequest._
-import org.apache.kafka.common.requests.{AlterConfigsRequest, ApiError, DescribeConfigsResponse, Resource, ResourceType}
+import org.apache.kafka.common.requests._
```
We don't want to do this.

ijuma commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -194,6 +197,64 @@ class AdminManager(val config: KafkaConfig,
+          "," * (oldNumPartitions) + assignments.map(_.mkString(":")).mkString(",")
+        }
+
+        AdminUtils.addPartitions(zkUtils, topic, newPartition.totalCount, reassignmentStr, validateOnly=validateOnly)
```
Nit: we add spaces before and after `=`.

ijuma commented on core/src/main/scala/kafka/server/AdminManager.scala:
```diff
@@ -194,6 +197,64 @@ class AdminManager(val config: KafkaConfig,
+      topicPurgatory.tryCompleteElseWatch(delayedCreate, delayedCreateKeys)
+    } else {
+      callback(done)
+    }
```
Is there a reason why we didn't follow the same approach as for `createTopics`?

ijuma commented on core/src/main/scala/kafka/server/DelayedCreatePartitions.scala:
```diff
@@ -0,0 +1,93 @@
+      return forceComplete()
+    } else {
+      false
+    }
```
Seems like the 5 lines above can be expressed as:

```scala
waitingFor.isEmpty && forceComplete()
```

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1333,6 +1335,40 @@ class KafkaApis(val requestChannel: RequestChannel,
+  def handleCreatePartitionsRequest(request: RequestChannel.Request): Unit = {
+    val alterPartitionCountsRequest = request.body[CreatePartitionsRequest]
+    var valid: Map[String, NewPartitions] = null
+    var errors: Map[String, ApiError] = null
```
There's no real reason for declaring these vars. `if/else` in Scala are expressions, so they can be assigned as vals. But I wonder why not follow the approach in `createTopics`. All things being equal, it's better to be consistent. If there are good reasons to deviate, then it's fine to do so.

ijuma commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -360,6 +361,168 @@ class AdminClientIntegrationTest extends KafkaServerTestHarness with Logging {
+    TestUtils.createTopic(zkUtils, topic1, 1, 1, servers, new Properties)
+
+    // assert that the topic has 1 partition
+    assertEquals(1, client.describeTopics(Set(topic1).asJavaCollection).values.get(topic1).get.partitions.size)
```
Nit: isn't `asJava` (instead of `asJavaCollection`) good enough here?

ijuma commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -360,6 +361,168 @@ class AdminClientIntegrationTest extends KafkaServerTestHarness with Logging {
+    val validateOnly = new CreatePartitionsOptions().validateOnly(true)
+
+    // assert that a validateOnly request doesn't increase the number of partitions
+    var alterResult = client.createPartitions(Map(topic1 -> NewPartitions.increaseTo(new Integer(2))).asJava, validateOnly)
```
Nit: a few long lines in this test. Also, no need for `new Integer` given auto-boxing.

ijuma commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -360,6 +361,168 @@ class AdminClientIntegrationTest extends KafkaServerTestHarness with Logging {
+      fail("Expect InvalidPartitionsException when newCount == oldCount")
+    } catch {
+      case e: ExecutionException =>
+        assertTrue(e.getCause.isInstanceOf[InvalidPartitionsException])
```
How do we detect a no-op? If we provide a way to do that, then one doesn't have to `describe` to know if the operation succeeded. Another option is that a no-op doesn't return an error. Pros and cons to that, so not sure.

tombentley replied:
The API is to create partitions, so its simply an error if the increment is 0 (or negative). See `AdminManager.createPartitions()`

ijuma replied:
I think it's important to think about how users will use the API. A common case is that someone wants to retry until the number of partitions is 10. Let's assume that the first call times out, but actually succeeds. On the second call, the user gets an error even though the desired result was achieved. How do you envisage the user dealing with this? Do they have to `describeTopics` to verify? It would be convenient to be able to handle this case without the `describe`, I think.

ijuma replied:
Btw, what I'm suggesting is quite common for idempotent APIs.

tombentley replied:
Well that would be OK in the case that no replica assignment is given, but if replica assignments are given we ought to return an error (because otherwise the user will get the impression that those assignments are in effect when in fact they might not be). But then you have an inconsistency in the API between the cases where assignments are given or not.

ijuma commented on core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala:
```diff
@@ -360,6 +361,168 @@ class AdminClientIntegrationTest extends KafkaServerTestHarness with Logging {
+    assertEquals(2, actualPartitions.size)
+
+    // now try creating a new partition (with assignments), to bring the total to 3 partitions
+    alterResult = client.createPartitions(Map(topic1 -> NewPartitions.increaseTo(new Integer(3), asList(asList(new Integer(0))))).asJava)
```
It would be good if the assignment was a little more complex.

ijuma reviewed (commented):
A few more comments. Almost finished with the review.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -1799,4 +1801,42 @@ void handleFailure(Throwable throwable) {
+
+    public CreatePartitionsResult createPartitions(Map<String, NewPartitions> newPartitions, final CreatePartitionsOptions options) {
+        final Map<String, KafkaFutureImpl<Void>> futures = new HashMap<>(newPartitions.size());
+        for (String topic: newPartitions.keySet()) {
```
Nit: space before `:`

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/CreatePartitionsRequest.java:
```diff
@@ -0,0 +1,215 @@
+
+public class CreatePartitionsRequest extends AbstractRequest {
+
+    private static final String TOPIC_PARTITION_COUNT_KEY_NAME = "topic_partition_count";
```
Maybe this should just be `topic_partitions` since it's composed of a sequence of `topic` and `newPartitions` (if the inner fields don't mention `count`, it's a bit weird for the outer one to do it).

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/CreatePartitionsRequest.java:
```diff
@@ -0,0 +1,215 @@
+                                    new Field(COUNT_KEY_NAME, INT32, "The new partition count."),
+                                    new Field(ASSIGNMENT_KEY_NAME, ArrayOf.nullable(new ArrayOf(INT32)), "The assigned brokers")
+                            )))),
+                    "List of topic and the corresponding partition count"),
```
Maybe "and the corresponding new partitions"

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/CreatePartitionsRequest.java:
```diff
@@ -0,0 +1,215 @@
+        return new Schema[]{CREATE_PARTITIONS_REQUEST_V0};
+    }
+
+    private final Set<String> duplicates;
```
Worth adding a comment explaining this.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/CreatePartitionsRequest.java:
```diff
@@ -0,0 +1,215 @@
+            NewPartitions dupe = counts.put(topic, newPartition);
+            if (dupe != null) {
+                if (dupes.isEmpty()) {
+                    dupes = new HashSet<>();
```
This isn't worth it, I don't think. Assigning a `new HashSet<>(0)` instead of `emptySet` is probably fine considering all the other work happening in this method.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/CreatePartitionsRequest.java:
```diff
@@ -0,0 +1,215 @@
+            String topic = topicPartitionCountStruct.get(TOPIC_NAME);
+            Struct partitionCountStruct = topicPartitionCountStruct.getStruct(NEW_PARTITIONS_KEY_NAME);
+            int count = partitionCountStruct.getInt(COUNT_KEY_NAME);
+            Object[] outerArray = partitionCountStruct.getArray(ASSIGNMENT_KEY_NAME);
```
`assignmentsArray` maybe?

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/CreatePartitionsRequest.java:
```diff
@@ -0,0 +1,215 @@
+            NewPartitions newPartition;
+            if (outerArray != null) {
+                List<List<Integer>> assignments = new ArrayList(outerArray.length);
+                for (Object inner : outerArray) {
```
`replicas` instead of `inner`?

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/CreatePartitionsRequest.java:
```diff
@@ -0,0 +1,215 @@
+                    i++;
+                }
+            } else {
+                x = null;
```
Just assign `null` from the start?

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/CreatePartitionsRequest.java:
```diff
@@ -0,0 +1,215 @@
+            topicPartitionCountStruct.set(NEW_PARTITIONS_KEY_NAME, partitionCountStruct);
+            topicPartitionsList.add(topicPartitionCountStruct);
+        }
+        struct.set(TOPIC_PARTITION_COUNT_KEY_NAME, topicPartitionsList.toArray(new Object[topicPartitionsList.size()]));
```
passing `new Object[0]` is the fastest and most concise way.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/CreatePartitionsRequest.java:
```diff
@@ -0,0 +1,215 @@
+                x = new Object[count.assignments().size()][];
+                int i = 0;
+                for (List<Integer> partitionAssignment : count.assignments()) {
+                    x[i] = partitionAssignment.toArray(new Object[partitionAssignment.size()]);
```
passing new Object[0] is the fastest and most concise way.

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/CreatePartitionsRequest.java:
```diff
@@ -0,0 +1,215 @@
+            NewPartitions count = topicPartitionCount.getValue();
+            Struct partitionCountStruct = topicPartitionCountStruct.instance(NEW_PARTITIONS_KEY_NAME);
+            partitionCountStruct.set(COUNT_KEY_NAME, count.totalCount());
+            Object[][] x;
```
x -> assignments?

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/CreatePartitionsResponse.java:
```diff
@@ -0,0 +1,105 @@
+
+public class CreatePartitionsResponse extends AbstractResponse {
+
+    private static final String TOPIC_ERRORS_KEY_NAME = "topic_partition_count_error";
```
Name can be simply `topic_errors` like the key name?

ijuma commented on clients/src/main/java/org/apache/kafka/common/requests/CreatePartitionsResponse.java:
```diff
@@ -0,0 +1,105 @@
+                            TOPIC_NAME,
+                            ERROR_CODE,
+                            ERROR_MESSAGE
+                    )), "Topic partition count results")
```
"Per topic results for the create partitions request"?

ijuma reviewed (commented):
OK, completed the review.

tombentley committed:
Fix more review comments

tombentley committed:
Factor out AdminUtils.parseManualReplicaAssignment, validateReplicaAssignment

tombentley committed:
Add a test

tombentley committed:
Use same purgatory logic as createTopics()

tombentley commented:
@ijuma I think I've addressed all your review comments except for the question about what the noop behaviour should be. Thinking about that, I guess it would be possible to validate that any passed assignments were identical to those already present, and return OK if they were.

ijuma commented:
Thanks @tombentley. I'll take a pass and if everything looks reasonable, I'll merge. We can file a JIRA for any potential follow-ups (including the noop behaviour). We can then discuss a bit more with a bit more time.

tombentley commented:
retest this please

asfgit commented:
SUCCESS 
 7808 tests run, 5 skipped, 0 failed.
--none--

tombentley commented:
@ijuma ftr I have a patch for the noop case, but I'm happy to do a follow up JIRA for that.

ijuma reviewed (approved):
Thanks for the updates, LGTM. I have some improvement suggestions, but I'll post them as a follow-up PR.

ijuma commented:
Seems like this PR added a flaky test:

https://builds.apache.org/blue/organizations/jenkins/kafka-trunk-jdk7/detail/kafka-trunk-jdk7/2795/tests

tombentley commented:
OK, let me look at it
